### ISSUE TITLE
needless option take improvement

### ISSUE THREAD

**Issue by InfRandomness** [rust-lang/rust-clippy#8740]
### What it does

The lint already exists but it needs an improvement in the way it detects temporaries : 
Currently, it matches a DefId against `&["core", "option", "Option", "as_ref"]`, while this works, this can give false negatives on temporary bindings.
The way we want that to work instead if described here : https://github.com/rust-lang/rust-clippy/pull/8665#issuecomment-1103939951

Upstream PR introducing the initial lint : https://github.com/rust-lang/rust-clippy/pull/8665

### Lint Name

needless_option_take

### Category

complexity

### Example

```rust
<code>
```

Could be written as:

```rust
<code>
```
